types/vote_test.go

```diff
@@ -336,38 +335,51 @@ func TestVoteString(t *testing.T) {
 	}
 }
 
-func TestVoteValidateBasic(t *testing.T) {
+func TestVoteValidation(t *testing.T) {
 	privVal := NewMockPV()
 
 	testCases := []struct {
-		testName     string
-		malleateVote func(*Vote)
-		expectErr    bool
+		testName                 string
+		malleateVote             func(*Vote)
+		expectBasicErr           bool
+		expectExtErrForPrevote   bool
+		expectExtErrForPrecommit bool
 	}{
-		{"Good Vote", func(v *Vote) {}, false},
-		{"Negative Height", func(v *Vote) { v.Height = -1 }, true},
-		{"Negative Round", func(v *Vote) { v.Round = -1 }, true},
+		{"Good Vote", func(v *Vote) {}, false, false, false},
+		{"Negative Height", func(v *Vote) { v.Height = -1 }, true, true, true},
+		{"Negative Round", func(v *Vote) { v.Round = -1 }, true, true, true},
 		{"Invalid BlockID", func(v *Vote) {
 			v.BlockID = BlockID{[]byte{1, 2, 3}, PartSetHeader{111, []byte("blockparts")}}
-		}, true},
-		{"Invalid Address", func(v *Vote) { v.ValidatorAddress = make([]byte, 1) }, true},
-		{"Invalid ValidatorIndex", func(v *Vote) { v.ValidatorIndex = -1 }, true},
-		{"Invalid Signature", func(v *Vote) { v.Signature = nil }, true},
-		{"Too big Signature", func(v *Vote) { v.Signature = make([]byte, MaxSignatureSize+1) }, true},
+		}, true, true, true},
+		{"Invalid Address", func(v *Vote) { v.ValidatorAddress = make([]byte, 1) }, true, true, true},
+		{"Invalid ValidatorIndex", func(v *Vote) { v.ValidatorIndex = -1 }, true, true, true},
+		{"Invalid Signature", func(v *Vote) { v.Signature = nil }, true, true, true},
+		{"Too big Signature", func(v *Vote) { v.Signature = make([]byte, MaxSignatureSize+1) }, true, true, true},
+		{"Vote extension present", func(v *Vote) { v.Extension = []byte("extension") }, false, true, false},
+		{"Missing vote extension signature", func(v *Vote) { v.ExtensionSignature = nil }, false, false, true},
 	}
 	for _, tc := range testCases {
 		tc := tc
 		t.Run(tc.testName, func(t *testing.T) {
 			ctx, cancel := context.WithCancel(context.Background())
 			defer cancel()
 
-			vote := examplePrecommit(t)
-			v := vote.ToProto()
-			err := privVal.SignVote(ctx, "test_chain_id", v)
-			vote.Signature = v.Signature
-			require.NoError(t, err)
-			tc.malleateVote(vote)
-			assert.Equal(t, tc.expectErr, vote.ValidateBasic() != nil, "Validate Basic had an unexpected result")
+			votes := []*Vote{examplePrevote(t), examplePrecommit(t)}
+			for i, vote := range votes {
+				v := vote.ToProto()
+				err := privVal.SignVote(ctx, "test_chain_id", v)
+				vote.Signature = v.Signature
+				require.NoError(t, err)
+				tc.malleateVote(vote)
+				// ValidateBasic errors should be consistent across prevotes and precommits
+				assert.Equal(t, tc.expectBasicErr, vote.ValidateBasic() != nil, "ValidateBasic had an unexpected result")
```

Comment:
Is there some reason these are `assert` instead of `require`?

Reply:
It seems as though different authors had different conventions throughout the history of the codebase. Do you have a preference for one of them?

Reply:
The answer is complicated. 

With t.Error (analogous to assert) and t.Fatal (analogous to require), I prefer Error where practical so that the test can keep providing useful feedback while it's running anyway. When I write tests with the standard handlers I usually only Fatal for cases where it doesn't make sense to continue. 

But with testify, I default to "require" (the fatal form) because it obviates the need to check the result and explicitly continue (which defeats the point of the assertion). Since the standard case already has the condition it's not a big addition. 

If it were up to me we wouldn't use either, but on the above rationale I usually default "require". YMMV.